src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -1002,12 +1002,20 @@ def query(self, filters, options):
             snapshots = zfs.snapshots_serialized(**kwargs)
 
         # FIXME: awful performance with hundreds/thousands of snapshots
+        select = options.pop('select', None)
         result = filter_list(snapshots, filters, options)
 
-        if isinstance(result, list):
-            result = self.middleware.call_sync('zettarepl.annotate_snapshots', result)
-        elif isinstance(result, dict):
-            result = self.middleware.call_sync('zettarepl.annotate_snapshots', [result])[0]
+        if not select or 'retention' in select:
+            if isinstance(result, list):
+                result = self.middleware.call_sync('zettarepl.annotate_snapshots', result)
+            elif isinstance(result, dict):
+                result = self.middleware.call_sync('zettarepl.annotate_snapshots', [result])[0]
+
+        if select:
+            if isinstance(result, list):
+                result = [{k: v for k, v in item.items() if k in select} for item in result]
```

Comment:
Why don't we do this before doing `filter_list` ? If a user has thousands of snapshots, we would be iterating twice ?

Reply:
Do what exactly?